dhall/src/phase/normalize.rs

```diff
@@ -372,6 +372,59 @@ enum Ret<'a> {
     Expr(ExprF<Thunk, X>),
 }
 
+/// Performs an outer join of two HashMaps.
+///
+/// # Arguments
+///
+/// * `ft` - Will convert the values of the first map
+///          into the target value.
+///
+/// * `fu` - Will convert the values of the second map
+///          into the target value.
+///
+/// * `ftu` - Will convert the key and values from both maps
+///           into the target type.
+///
+/// # Description
+///
+/// If the key is present in both maps then the final value for
+/// that key is computed via the `ftu` function. Otherwise, the
+/// final value will be calculated by either the `ft` or `fu` value
+/// depending on which map the key is present in.
+///
+/// The final map will contain all keys from the two input maps with
+/// also values computed as per above.
+pub(crate) fn outer_join<K, T, U, V>(
+    mut ft: impl FnMut(&T) -> V,
+    mut fu: impl FnMut(&U) -> V,
+    mut ftu: impl FnMut(&K, &T, &U) -> V,
+    map1: &HashMap<K, T>,
+    map2: &HashMap<K, U>,
+) -> HashMap<K, V>
+where
+    K: std::hash::Hash + Eq + Clone,
+{
+    let mut kus = HashMap::new();
```

Comment:
Should this be named kvs, since it contains values of type V ?

Reply:
Good call, did some renaming :ok_hand: